pkg/madmin/config-commands.go

```diff
@@ -75,6 +78,17 @@ func (adm *AdminClient) SetConfig(config io.Reader) (r SetConfigResult, err erro
 		return r, err
 	}
 
+	// Check if read configBtyes is in json format
+	err = json.Unmarshal(configBytes, &configMap)
+	if err != nil {
+		return r, errors.New("Invalid JSON format: " + err.Error())
+	}
+
+	// Make sure "version" key exists
+	if _, ok := configMap["version"]; !ok {
```

Comment:
This is already validated by quick.Config when the config structure is being saved. The main part of validation here is to validate all fields basically equal the same amount of steps involved when starting a server. 

- Being able to unmarshal the incoming content into a valid struct
- Once unmarshalled an attempt is made to 
   - Check if access credentials are correct?
   - Check if domain is correct? 
   - Check if storageClass is correct?
   - Check if cache is correct?
   - An attempt is made to initialize the notification subsystem i.e validate all the notification targets if one of them
     fails return error to the client. (Webhook target is okay to be not available if it's not available it is not initialized)

Reply:
@harshavardhana,
This is the client side check on the configuration json file, which could be partial or complete, when user runs a command to update/set Minio configuration :
```
    $ cat <config.json> | mc admin --insecure config set <alias>
```
We decided to check **only** the **existence** of certain keys, only `"version"` at this point, but not their **correctness**. That is because  we expect a generic json file 
```
    var configMap map[string]interface{}
```
as client side doesn't have the configuration structure defined. The struct info is only available on the server side.

Anyways we'll go through extensive input validation and checks on the server side to make sure the json structure, partial or complete, has **valid** key names and **correct** types for values. 

So, the thought is that only `"version"` key must be included whenever a partial or complete configuration json file is provided. 
Also we do not need to check the `accessKey` and `secretKey` on the client side, since it is checked at the server side if corresponding environment variables are defined for them.

I am currently working with @vadmeste, who will work on the server side, to make sure we are all on the same page.

Reply:
> We decided to check only the existence of certain keys, only "version" at this point, but not their correctness. That is because we expect a generic json file

Yes even here we should simply use existing config structure, JSON using configMap is not correct this looses all types and we have removed the real requirement of validating the config.json

The client can set any structure and send - if they are enabled we should validate them before writing to the disk without these checks we can in-fact end up taking down the server.

Reply:
... but server side will check and validate before writing. So, it will never be written onto the disk until validation is complete.
The reason we wanted to do correctness check on the server side is the fact that it didn't make sense to `get` and validate the values on the client side while server side already has all the information.

Reply:
It is not doing that right now @ebozduman, this PR should address that. Currently, it simply parses and writes to disk even with incorrect values.

Reply:
```diff
git diff
diff --git a/pkg/madmin/config-commands.go b/pkg/madmin/config-commands.go
index cf120c68..121c2f9b 100644
--- a/pkg/madmin/config-commands.go
+++ b/pkg/madmin/config-commands.go
@@ -23,6 +23,8 @@ import (
        "io"
        "io/ioutil"
        "net/http"
+
+       "github.com/minio/minio/pkg/quick"
 )
 
 // NodeSummary - represents the result of an operation part of
@@ -63,15 +65,42 @@ func (adm *AdminClient) GetConfig() ([]byte, error) {
        return ioutil.ReadAll(resp.Body)
 }
 
+const maxConfigJSONSize = 256 * 1024 // 256KiB
+
 // SetConfig - set config supplied as config.json for the setup.
 func (adm *AdminClient) SetConfig(config io.Reader) (r SetConfigResult, err error) {
        if !adm.secure { // No TLS?
                return r, fmt.Errorf("credentials/configuration cannot be updated over an insecure connection")
        }
 
-       // Read config bytes to calculate MD5, SHA256 and content length.
-       configBytes, err := ioutil.ReadAll(config)
-       if err != nil {
+       // Read configuration bytes from request body.
+       configBuf := make([]byte, maxConfigJSONSize+1)
+       n, err := io.ReadFull(config, configBuf)
+       if err == nil {
+               return r, fmt.Errorf("too large file")
+       }
+       if err != io.ErrUnexpectedEOF {
+               return r, err
+       }
+
+       configBytes := configBuf[:n]
+
+       type configVersion struct {
+               Version string `json:"version"`
+       }
+       var cfg configVersion
+
+       if err = json.Unmarshal(configBytes, &cfg); err != nil {
+               return r, err
+       }
+
+       if _, err = quick.NewConfig(cfg, nil); err != nil {
+               return r, err
+       }
+
+       // Validate JSON provided in the request body: check the
+       // client has not sent JSON objects with duplicate keys.
+       if err = quick.CheckDuplicateKeys(string(configBytes)); err != nil {
                return r, err
        }
```